### PR TITLE
Allow user with empty email if email-domain set to wildcard

### PR DESCRIPTION
**Pull request by urbanpawel90**
## Description

Issue #769 possible fix. I assumed that if we have `email-domain` set to `*` we will also validate users with empty email address.

## Motivation and Context

Possible solution to #769 

## How Has This Been Tested?

Unit test was added.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: validator.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func newValidatorImpl(domains []string, usersFile string,
 	}
 
 	validator := func(email string) (valid bool) {
-		if email == "" {
+		if !allowAll && email == "" {
```
Comment (JoelSpeed) [inline validator.go:90]: I'd be tempted to make this more explicit, and move the block from below to the top of here

Eg 

```
validator := func(email string) (valid bool) {
  if allowAll {
    return true
  }

  // rest of function as is, removing the same if statement from below
```

**Review by JoelSpeed — COMMENTED** [review]
Looks good! Added a couple of suggestions.

The only thing I'm considering at this point is whether we should be treating this as a bug and meaning it can be released as a patch, or whether we should be treating this as a breaking change 🤔 Since it technically makes things less secure, it could be interpreted as the latter

File: validator.go (hunk 1 above)
Comment (urbanpawel90) [reply validator.go:90]: Will introduce that change in a minute

Comment (urbanpawel90) [reply validator.go]: Suggestion applied

**Comment by NickMeves** [comment]
> Looks good! Added a couple of suggestions.
> 
> The only thing I'm considering at this point is whether we should be treating this as a bug and meaning it can be released as a patch, or whether we should be treating this as a breaking change 🤔 Since it technically makes things less secure, it could be interpreted as the latter

I'd lean toward this needing to go into at least a minor release -- potentially a major release due to the aspect you mentioned that we would now authorize tokens in some environments that we would previously reject.

As I mentioned here https://github.com/oauth2-proxy/oauth2-proxy/issues/769#issuecomment-689100352 (before seeing this PR 🤦 ) -- a definite breaking change we should look into is the canonical treatment of email and forced hacks to create an email on the session when leaving it blank to match the original token claim is likely ideal.

File: validator.go [hunk 2]
```diff
@@ -87,6 +87,9 @@ func newValidatorImpl(domains []string, usersFile string,
 	}
 
 	validator := func(email string) (valid bool) {
+		if allowAll {
+			valid = true
```
Comment (JoelSpeed) [inline validator.go:91]: We can just return here, no need to do anything more
```suggestion
			return true
```

**Review by JoelSpeed — COMMENTED** [review]
I was leaning to a minor release, though I saw @urbanpawel90 made a suggestion of adding a flag that allows empty emails, this would definitely be the safe way to go, though I'm reluctant to add yet another option. Since this only affects users using the wildcard option, I would expect most of them would expect an empty email to be allowed anyway, I'm tempted to leave this open for a couple of weeks to let it soak and see if anyone else chips in

**Comment by urbanpawel90** [comment]
> I was leaning to a minor release, though I saw @urbanpawel90 made a suggestion of adding a flag that allows empty emails, this would definitely be the safe way to go, though I'm reluctant to add yet another option. Since this only affects users using the wildcard option, I would expect most of them would expect an empty email to be allowed anyway, I'm tempted to leave this open for a couple of weeks to let it soak and see if anyone else chips in

So maybe let's agree to some due date after which we could proceed with the PR? The end of September would be ok or you want to wait longer?

**Comment by JoelSpeed** [comment]
Let's do this properly and consider it breaking. I've discussed with @NickMeves and we have a number of breaking PRs queueing up again now, so let's start merging those and get ready for a v7 release

**Comment by NickMeves** [comment]
I think the Github provider has some logic where it might've assumed a token without an email would always be denied. I'll need to circle back and confirm as all the provider authorization refactors are wrapping up.

But I think it was this logic where a nil error is returns even though it is a scenario where an error definitely should be passed for the `!ok` case: https://github.com/oauth2-proxy/oauth2-proxy/blob/8bed7aafbdb2cb6241b9655e983dd54d54b50531/providers/github.go#L407

I haven't done a thorough check of other providers to see where potentially insecure logic was masked because it didn't set an email on the session in the end.

**Comment by JoelSpeed** [comment]
@NickMeves I think that's the only thing blocking this PR right now right? Otherwise we are happy to continue? 

@urbanpawel90 Would you be able to rebase this and make sure that a note is added to the breaking changes explaining that empty emails will now be allowed when a wildcard is supplied

**Comment by NickMeves** [comment]
> @NickMeves I think that's the only thing blocking this PR right now right? Otherwise we are happy to continue?
> 
> @urbanpawel90 Would you be able to rebase this and make sure that a note is added to the breaking changes explaining that empty emails will now be allowed when a wildcard is supplied

I would lean toward holding off on this one until we get a better feel for how all the providers behave after the refactoring activity to deprecate `GetEmailAddress` in favor of `EnrichSessionState`.

There was a lot of hacky AuthZ logic jammed into that method and we've already found one instance (Github) where the current implementation of this flag requiring an email on the session prevented a security bug.

**Pull request closed without merging**